server/shadow/shadow_server.c

```diff
@@ -510,16 +536,60 @@ int shadow_server_start(rdpShadowServer* server)
 		return -1;
 	}
 
-	if (!server->ipcSocket)
-		status = server->listener->Open(server->listener, NULL, (UINT16)server->port);
-	else
-		status = server->listener->OpenLocal(server->listener, server->ipcSocket);
+	/* Bind magic:
+	 *
+	 * emtpy                 ... bind TCP all
```

Comment:
TCP/IPv4 only, at least with the current implementation on Linux

Reply:
no, why?

Reply:
only limitation is you can only bind 1 interface with the current API

Reply:
I'm not exactly sure why — I think _something_ is causing the shadow server to only bind the first address returned by `getaddrinfo`, but I'm not sure why — my interpretation of `listener.c` https://github.com/FreeRDP/FreeRDP/blob/ce03f5eac59a26a667a578532ab783699b65bc08/libfreerdp/core/listener.c#L74 suggests that it would bind all of them.


```
[linus@bringtnix:~/projects/freerdp]$ ./server/shadow/freerdp-shadow-cli
[09:02:02:439] [30228:30228] [INFO][com.freerdp.server.shadow.x11] - X11 Extensions: XFixes: 1 Xinerama: 1 XDamage: 0 XShm: 0
[09:02:02:442] [30228:30228] [INFO][com.freerdp.core.listener] - Listening on [0.0.0.0]:3389

<in another terminal>
[linus@bringtnix:~]$ nc -v6 localhost 3389
nc: connect to localhost port 3389 (tcp) failed: Connection refused

[linus@bringtnix:~]$ nc -v localhost 3389
nc: connect to localhost port 3389 (tcp) failed: Connection refused
Connection to localhost 3389 port [tcp/ms-wbt-server] succeeded!
```

Reply:
@lheckemann check out `MAX_LISTENER_HANDLES` which only allows 5 listeners.

Reply:
Sure, but it's only listening on the one address, so I don't think that limit is what's stopping it from listening on IPv6.

Reply:
@lheckemann well, if I start with `/bind-address:[::]` or `/bind-address:[::1]` it listens ;)

Reply:
Right, but "TCP on all IPv4 addresses" isn't the same thing as "TCP on all IP addresses", and the latter is what I'd expect given no arguments :/ that said, I wouldn't consider it critical.

Reply:
@lheckemann added some code for the tcp path to allow multiple bind addresses.